src/recipe-selector/recipe-selector.ts

```diff
@@ -11,14 +11,54 @@
 import {Recipe} from '../runtime/recipe/recipe.js';
 import {Dictionary} from '../runtime/hot.js';
 
+// One entry in the lookup table, i.e. one trigger and the recipe it invokes.
+export class Match {
+  constructor(readonly trigger: [string, string][], readonly recipe: Recipe) {}
+
+  // The definition of a matching trigger is embodied in this function.
+  // This first implementation matches if every key-value pair in the trigger
+  // is present in the request. The request may contain other pairs, and the
+  // order of the matching pairs need not be the same as the order in the trigger. 
+  matches(request: [string, string][]): boolean {
+    if (this.trigger.length > request.length) return false;
+    for (const pair of this.trigger) {
+      const found = request.findIndex(r => r[0] === pair[0] && r[1] === pair[1]);
+      if (found === -1) return false;
+    }
+    return true;
+  }
+}
+
 export class RecipeSelector {
+    private _table: Match[] = [];
+
+    // For testing only
+    get table() {
+      return this._table;
+    }
+
     // Only recipes with “@trigger” annotations get included in the lookup table.
-    // Other recipes are silently ignored.
-    // Specific semantics of overlapping triggers are still TBD.
+    // Other recipes are silently ignored. This relies on the triggers array
+    // being always present but empty if there are no triggers.
+    // The order of the Matches in the lookup table preserves the order of the
+    // Recipes in the constructor argument, and within each recipe the order of
+    // the triggers in the triggers field of the recipe.
```

Comment:
RecipeSelector could instead take Manifest as ctor parameter and own the logic of which recipes should be used.

Reply:
There actually isn't much logic. The triggers are part of the recipes, initialized to an empty array and filled if any triggers are encountered when the manifest is parsed. The parsing does already preserves both orders, and I've included a test to ensure that this remains the case. So this code just preserves order when building the lookup table, and ignores recipes with empty trigger arrays. So unless the parsing changes to do something to the order, there really isn't anything to be gained by taking the manifest on the constructor.